### PR TITLE
Boundary distance model 

### PR DESCRIPTION
**Pull request by DaoudPiracha**
### Major Features:
- **New Model Class**: 
  - Added a derived class `SegmenterBoundaryDistance` in `ssl4rs/models/classif/base.py`. This implements a pixel-wise regression model to train on a new mask, `distance_from_boundary` (defined in `ssl4rs/data/transforms/distance_from_boundary.py`)

- **New Loss**: MaskedMSELoss 
- **New Metric**: MaskedMeanSquaredError

- **New Transforms**:
  - Added the following transforms:
    - `distance_from_boundary` in `ssl4rs/data/transforms/distance_from_boundary.py` to generate distance masks from boundary masks. This mask assigns real values within a field boundary in the range [0,1] based on their proximity to the nearest boundary point.
    - `composite_bands` in `ssl4rs/data/transforms/composite_bands.py`.
    - `grab_image_from_stack` in `ssl4rs/data/transforms/grab_image_from_stack.py`.

- **Bug Fix**:
  - Resolved an issue in the custom collate function that was causing batch images to not load properly. *(Note: This fix was part of an older commit. I will either, move htat commit to a seperate PR or revisit and refine this comment in a future PR edit.)*

### Minor Changes:

- **Changes to Field Boundary (0/1) Mask Generation**:

  - Added a padding step for increased safety in `generate_boundary_distance_mask`, specifically for cases where a contour is connected to the edge of the image array.
  - Introduced additional assertions in `generate_boundary_distance_mask` located in `ssl4rs/data/transforms/boundary.py`.

- **Configuration Update**:
  - Updated `ssl4rs/utils/config.py`:
  
  
    ```python
    "runtime_hash": hashlib.sha1(str(curr_time).encode(), usedforsecurity=False).hexdigest(),
    ```
    changed to:
    ```python
    "runtime_hash": hashlib.sha1(str(curr_time).encode()).hexdigest(),
    ```
    - This change was made because the second hash wasn't being interpreted correctly, potentially due to a versioning issue, and was causing the code to crash. The removal is a temporary fix; a more thorough and possibly easier solution may exist.

### Input Needed:
- **MaskedMSELoss and Metric MaskedMeanSquaredError**:
  - These are currently implemented directly in `ssl4rs/models/classif/base.py`. Since this is a merge to the master branch, it might be worth considering moving them to more appropriate locations, though it’s unclear where that should be at the moment.
  - Also, I’m uncertain how to add the hook so Hydra can directly import these components. Guidance would be appreciated.

---

### Checklist Before Submission:
- [ x] Ensure the **PR title is self-explanatory** and the **description concisely explains the PR**.
- [ ] Confirm that your **PR does only one thing**, avoiding the bundling of different changes.
Note: this is currently bundling, will update in draft 2.0. 
- [ ] List all **breaking changes** introduced by this pull request.
- [ ] Run all pre-commit hooks manually with the `make check` command.
- [ ] Test the PR locally with the `make test-all` command.

### CODE REVIEW

File: ssl4rs/data/datamodules/disa.py [hunk 1]
```diff
@@ -173,7 +174,6 @@ def test_dataloader(self) -> torch.utils.data.DataLoader:
         assert self.data_test is not None, "parser unavailable, call `setup()` first!"
         return self._create_dataloader(self.data_test, subset_type="test")
 
-
```
Comment (plstcharles) [inline ssl4rs/data/datamodules/disa.py:176]: removing the extra newline here causes a formatting error

File: ssl4rs/data/datamodules/disa.py [hunk 2]
```diff
@@ -246,6 +251,15 @@ def convert_deeplake_tensors_to_pytorch_tensors(
...
 
+# todo: maybe change so this is not inplace?
+def convert_4_band_to_3_band(
+    batch: ssl4rs.data.BatchDictType,
+) -> ssl4rs.data.BatchDictType:
+    """Transform converts RGB+NIR into a 3-band composite suitable for pretrained backbones"""
+    img_data = batch['image_data']
+    img_transform = ssl4rs.data.transforms.composite_bands.Convert4BandTo3Band()
+    batch['image_data'] = img_transform(img_data)
+    return batch
```
Comment (plstcharles) [inline ssl4rs/data/datamodules/disa.py:262]: there are some formatting issues in this function, make sure to run the linter ("make check") to catch/fix them

File: ssl4rs/data/datamodules/disa.py [hunk 3]
```diff
@@ -268,6 +282,28 @@ def generate_field_boundary_mask(
...
+    batch: ssl4rs.data.BatchDictType,
+    output_distance_mask_name: str = "distance_from_boundary_mask",
+) -> ssl4rs.data.BatchDictType:
+    """Transform used in parser class to generate distance from field boundary (contour) masks."""
+    assert isinstance(batch, dict)
+    assert "field_mask" in batch
+    class_map = batch["field_mask"]
+    unannotated_mask = class_map == 0  # this is the real 'dontcare' which we reapply below
+    distance_mask = ssl4rs.data.transforms.distance_from_boundary.generate_boundary_distance_mask(
+    class_label_map = class_map,
```
Comment (plstcharles) [inline ssl4rs/data/datamodules/disa.py:295]: there's probably a tab alignment/formatting error here

File: ssl4rs/data/datamodules/disa.py [hunk 4]
```diff
@@ -268,6 +282,28 @@ def generate_field_boundary_mask(
...
+    target_class_label = 1,  # target the "positive" (field) class inside the binary mask
+    ignore_index = DataModule.metadata.dontcare_label,
+    pad_size = 1
```
Comment (plstcharles) [inline ssl4rs/data/datamodules/disa.py:298]: keyword args passed to a function would not have spaces around the '='

File: ssl4rs/data/metadata/disa.py [hunk 5]
```diff
@@ -168,7 +168,7 @@
     "field_geoms",  # since both polygon counts and point counts can vary across locations
     "image_order_ids",  # list of strings/identifiers
     "image_metadata",  # list of dicts with lots of more-or-less useful stuff in them
-    "image_data",  # array whose first dimension varies (it's the number of matched orders)
+    #"image_data",  # array whose first dimension varies (it's the number of matched orders)
```
Comment (plstcharles) [inline ssl4rs/data/metadata/disa.py:171]: not sure why this is getting commented here? (likely breaks code elsewhere, this needs to be changed)

File: ssl4rs/data/transforms/batch.py [hunk 6]
```diff
@@ -290,7 +290,6 @@ def get_batch_index(
     assert isinstance(batch_index, typing.Hashable), f"invalid batch index type: {type(batch_index)}"
     return batch_index
 
-
```
Comment (plstcharles) [inline ssl4rs/data/transforms/batch.py:293]: this module seems to contain changes that should not happen or do not need to happen; these should be eliminated

File: ssl4rs/data/transforms/boundary.py [hunk 7]
```diff
@@ -41,11 +42,36 @@ def generate_boundary_mask_from_class_label_map(
         else:
             return dontcare_mask.astype(np.int64) * ignore_index
     roi_mask = roi_mask.astype(np.uint8) * 255  # to prep for opencv input
+
+    roi_mask = np.pad(roi_mask, pad_size, mode='constant', constant_values=0)
+
+
+    if roi_mask.sum().sum() == 0:
+        raise ValueError("no fields detected in roi mask")
```
Comment (plstcharles) [inline ssl4rs/data/transforms/boundary.py:50]: if this function is not DISA-specific, it shouldn't say "no fields", and instead return something empty, or at least have a more generic error message

File: ssl4rs/data/transforms/boundary.py [hunk 8]
```diff
@@ -41,11 +42,36 @@ def generate_boundary_mask_from_class_label_map(
...
+    if roi_mask.dtype != np.uint8:
+        print("Converting image to uint8.")
+        roi_mask = roi_mask.astype(np.uint8)
+
+    # Converting image to grayscale if not already converted
+    if len(roi_mask.shape) == 3 and roi_mask.shape[2] == 3:
+        roi_mask = cv.cvtColor(roi_mask, cv.COLOR_BGR2GRAY)
+
+    if not roi_mask.flags['C_CONTIGUOUS']:
+        roi_mask = np.ascontiguousarray(roi_mask)
```
Comment (plstcharles) [inline ssl4rs/data/transforms/boundary.py:64]: using ascontiguousarray directly without the check would be cleaner/simpler

File: ssl4rs/data/transforms/distance_from_boundary.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
+import typing
+
+import cv2 as cv
+import numpy as np
+import torch
+
+
+def validate_inputs(class_label_map: np.ndarray, target_class_label: int, ignore_index: typing.Optional[int]):
```
Comment (plstcharles) [inline ssl4rs/data/transforms/distance_from_boundary.py:8]: if this validation function is only intended to be used inside this module, it should be private, so have a `_` prefix

File: ssl4rs/data/transforms/distance_from_boundary.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    each pixel has a distance value from the nearest boundary
+    Parameters
+    ----------
+    class_label_map
+    target_class_label
+    ignore_index
+    pad_size
+
+    Returns
+    -------
```
Comment (plstcharles) [inline ssl4rs/data/transforms/distance_from_boundary.py:33]: This docstring does not follow the Google docstring style used elsewhere

File: ssl4rs/data/transforms/distance_from_boundary.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        interior_indices = np.transpose(np.nonzero(contour_mask > 0))
+        distances_to_centroid = np.sqrt((interior_indices[:, 1] - cx) ** 2 + (interior_indices[:, 0] - cy) ** 2)
+        max_distance_to_centroid = distances_to_centroid.max()
+
+        # add distances from boundary and centroid
+        # todo: see if you want to divide instead or smth simimlar
+        # maybe 1 - (dist from bound/ dist from centroid)
+        if max_distance_to_centroid != 0:
+            combined_distances = (1 * distance_transform[interior_indices[:, 0], interior_indices[:, 1]]
+                                  + 0 * distances_to_centroid)
```
Comment (plstcharles) [inline ssl4rs/data/transforms/distance_from_boundary.py:106]: not sure why there's a "zero times anything" operation hidden in here?

File: ssl4rs/data/transforms/distance_from_boundary.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            max_combined_distance = combined_distances.max()
+            normalized_distances[interior_indices[:, 0], interior_indices[:, 1]] = (
+                    combined_distances / max_combined_distance
```
Comment (plstcharles) [inline ssl4rs/data/transforms/distance_from_boundary.py:109]: what is the meaning / intuition behind this ratio?

File: ssl4rs/models/classif/base.py [hunk 13]
```diff
@@ -367,3 +368,138 @@ def _render_and_log_samples(
...
+        preds, targets = outputs["preds"], outputs["targets"]
+        assert targets.ndim == 4 and targets.shape[0] == batch_size and targets.dtype == torch.float32
+        tensor_shape = targets.shape[2:]
+        assert preds.ndim == 4 and preds.shape == (batch_size, self.num_output_classes, *tensor_shape)
+
+        # we'll render the input tensors with a prediction mask and target mask side-by-side
+        outputs = []
+        for sample_idx, sample_id in zip(sample_idxs, sample_ids):
+            input_tensor = batch[self.input_key][sample_idx].cpu()
+            img_rgb = batch['location_preview_image'][sample_idx].cpu()
```
Comment (plstcharles) [inline ssl4rs/models/classif/base.py:492]: this makes the class DISA-specific, so many it should be in a disa subfolder or something

File: ssl4rs/models/classif/base.py [hunk 14]
```diff
@@ -367,3 +368,138 @@ def _render_and_log_samples(
...
+        self.mse_metric.update(valid_predictions, valid_targets)
+
+    def compute(self):
+        return self.mse_metric.compute()
+
+    def reset(self):
+        self.mse_metric.reset()
+
+
+class SegmenterBoundaryDistance(GenericSegmenter):
```
Comment (plstcharles) [inline ssl4rs/models/classif/base.py:410]: this class should be in a new (seperate) module, currently it's hidden in a place we don't expect to find it

File: ssl4rs/utils/config.py [hunk 15]
```diff
@@ -189,7 +189,7 @@ def get_runtime_tags(
         "git_hash": get_git_revision_hash(),
         "time_since_epoch": curr_time,
         "local_timestamp": time.strftime("%Y-%m-%d_%H-%M-%S", curr_local_time),
-        "runtime_hash": hashlib.sha1(str(curr_time).encode(), usedforsecurity=False).hexdigest(),
+        "runtime_hash": hashlib.sha1(str(curr_time).encode()).hexdigest(),
```
Comment (plstcharles) [inline ssl4rs/utils/config.py:192]: this should not be changing, it was there for a reason

File: tests/data/transforms/test_distance_from_boundary.py [hunk 16]
```diff
@@ -0,0 +1,135 @@
+import unittest
+import numpy as np
+import torch
+import matplotlib.pyplot as plt
+
+from ssl4rs.ssl4rs.data.transforms.distance_from_boundary import generate_boundary_distance_mask
+
+class TestGenerateBoundaryDistanceMask(unittest.TestCase):
```
Comment (plstcharles) [inline tests/data/transforms/test_distance_from_boundary.py:8]: the rest of the framework uses pytest, so this should use the same (instead of unittest)

File: tests/data/transforms/test_distance_from_boundary.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+            [2, 1, 0, 0, 0, 0, 0, 1, 2],
+            [2, 1, 1, 1, 1, 1, 1, 1, 2],
+            [2, 2, 2, 2, 2, 2, 2, 2, 2]
+        ], dtype=np.int32)
+        target_class_label = 1
+        expected_output = np.zeros_like(class_label_map)
+        output = generate_boundary_distance_mask(class_label_map, target_class_label, ignore_index=2)
+        self.visualize_result(output, 'Single Target Class Label')
+
+        # np.testing.assert_array_equal(output, expected_output)
```
Comment (plstcharles) [inline tests/data/transforms/test_distance_from_boundary.py:40]: not sure why the asserts would be commented out?

File: tests/data/transforms/test_distance_from_boundary.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+            [2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 2],
+            [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2],
+
+        ], dtype=np.int32)
+        target_class_label = 1
+        expected_output = np.zeros_like(class_label_map)
+        output = generate_boundary_distance_mask(class_label_map, target_class_label, ignore_index=2)
+        # self.visualize_result(output, 'Single Target Class Label')
+
+        # np.testing.assert_array_equal(output, expected_output)
```
Comment (plstcharles) [inline tests/data/transforms/test_distance_from_boundary.py:65]: not sure why the asserts would be commented out?

**Review by plstcharles — CHANGES_REQUESTED** [review]
there is a significant number of changes/cleanups that would be needed to merge this right now --- I am a bit more stringent on the code going into this repo, and it might be easier to port these changes into the private repo instead (we could discuss this later this week if needed)
